flume-ng-channels/flume-file-channel/src/test/java/org/apache/flume/channel/file/TestFileChannelErrorMetrics.java

```diff
@@ -0,0 +1,222 @@
+    assertEquals(2, channelCounter.getEventTakeAttemptCount());
+    assertEquals(1, channelCounter.getEventTakeSuccessCount());
+    assertEquals(1, channelCounter.getEventTakeErrorCount());
+  }
+
+  /**
+   * Test the FileChannelCounter.eventTakeErrorCount value if the data file
+   * contains an invalid record thus CorruptEventException is thrown during
+   * the take() operation.
+   * The first byte of the record (= the first byte of the file in this case)
+   * is the operation byte, changing it to an unexpected value will cause the
+   * CorruptEventException to be thrown.
+   */
+  @Test
+  public void testCorruptEventTaken() throws Exception {
+    FileChannel channel = createFileChannel(
+        Collections.singletonMap(FileChannelConfiguration.FSYNC_PER_TXN, "false"));
+    channel.start();
+
+    FileChannelCounter channelCounter = channel.getChannelCounter();
+
+    Transaction tx = channel.getTransaction();
+    tx.begin();
+    channel.put(EventBuilder.withBody("test".getBytes()));
+    tx.commit();
+    tx.close();
+
+    byte[] data = FileUtils.readFileToByteArray(new File(dataDirs[0], "log-1"));
+    data[0] = LogFile.OP_EOF; // change the first (operation) byte to unexpected value
+    FileUtils.writeByteArrayToFile(new File(dataDirs[0], "log-1"), data);
+
+    tx = channel.getTransaction();
+    tx.begin();
+
+    try {
+      channel.take();
+    } catch (Throwable t) {
+      // If fsyncPerTransaction is false then Log.get throws the CorruptEventException
+      // without wrapping it to IOException (which is the case when fsyncPerTransaciton is true)
+      // but in this case it is swallowed in FileBackedTransaction.doTake()
+      // The eventTakeErrorCount should be increased regardless of this.
+      Assert.fail("No exception should be thrown as fsyncPerTransaction is false");
+    }
+
+    assertEquals(1, channelCounter.getEventTakeAttemptCount());
+    assertEquals(0, channelCounter.getEventTakeSuccessCount());
+    assertEquals(1, channelCounter.getEventTakeErrorCount());
+  }
+
+  @Test
+  public void testCheckpointWriteErrorCount() throws Exception {
+    int checkpointInterval = 1000;
+    FileChannel channel = createFileChannel(Collections.singletonMap(
+        FileChannelConfiguration.CHECKPOINT_INTERVAL, String.valueOf(checkpointInterval)));
+    channel.start();
+
+    Transaction tx = channel.getTransaction();
+    tx.begin();
+    channel.put(EventBuilder.withBody("test".getBytes()));
+    tx.commit();
+    tx.close();
+
+    long beforeCheckpointWrite = System.currentTimeMillis();
+
+    // first checkpoint should be written successfully -> the counter should remain 0
+    Thread.sleep(checkpointInterval + 100);
+    assertEquals(0, channel.getChannelCounter().getCheckpointWriteErrorCount());
+    assertTrue(new File(checkpointDir, "checkpoint").lastModified() > beforeCheckpointWrite);
+
+    FileUtils.deleteDirectory(baseDir);
+    Thread.sleep(checkpointInterval + 100);
```

Comment:
Thank you for your suggestion. I created a custom class as we are on Java 7, but the logic itself is the same.